util/progress/progressui/printer.go

```diff
@@ -325,6 +222,252 @@ func (p *textMux) print(t *trace) {
+				}
+				tm = fmt.Sprintf(" %.1fs", dt)
+			}
+			fmt.Fprintf(p.w, "#%d DONE%s\n", v.index, tm)
+		}
+	}
+}
+
+func (text *textPrinter) printStatusUpdate(p *textMux, v *vertex, s *status) {
+	var bytes string
+	if s.Total != 0 {
+		bytes = fmt.Sprintf(" %.2f / %.2f", units.Bytes(s.Current), units.Bytes(s.Total))
+	} else if s.Current != 0 {
+		bytes = fmt.Sprintf(" %.2f", units.Bytes(s.Current))
+	}
+	var tm string
+	endTime := s.Timestamp
+	if s.Completed != nil {
+		endTime = *s.Completed
+	}
+	if s.Started != nil {
+		diff := endTime.Sub(*s.Started).Seconds()
+		if diff > 0.01 {
+			tm = fmt.Sprintf(" %.1fs", diff)
+		}
+	}
+	if s.Completed != nil {
+		tm += " done"
+	}
+	fmt.Fprintf(p.w, "#%d %s%s%s\n", v.index, s.ID, bytes, tm)
+}
+
+type jsonPrinter struct {
+	seen map[digest.Digest]struct{}
+	buf  SolveStatus
+}
+
+func (js *jsonPrinter) printVtx(p *textMux, t *trace, v *vertex) {
+	if js.seen == nil {
+		js.seen = make(map[digest.Digest]struct{})
+	}
+
+	if !p.notFirst {
+		if p.desc != "" {
+			js.buf.Vertex.Name = p.desc
+			js.print(p.w)
+		}
+		p.notFirst = true
+	}
+
+	js.buf.Digest = v.Digest
+	if _, ok := js.seen[v.Digest]; !ok {
+		js.buf.Vertex.Name = p.getVtxName(v)
+		js.buf.Vertex.Started = v.Started
+		js.seen[v.Digest] = struct{}{}
+	}
+
+	isOpenStatus := p.processStatusUpdates(v, func(s *status) {
+		js.buf.Statuses = append(js.buf.Statuses, VertexStatus{
+			ID:        s.VertexStatus.ID,
+			Name:      s.VertexStatus.Name,
+			Total:     s.VertexStatus.Total,
+			Current:   s.VertexStatus.Current,
+			Timestamp: s.VertexStatus.Timestamp,
+			Started:   s.VertexStatus.Started,
+			Completed: s.VertexStatus.Completed,
+		})
+	})
+
+	for _, w := range v.warnings[v.warningIdx:] {
+		js.buf.Warnings = append(js.buf.Warnings, VertexWarning{
+			Level:      w.Level,
+			Short:      w.Short,
+			Detail:     w.Detail,
+			URL:        w.URL,
+			SourceInfo: w.SourceInfo,
+			Range:      w.Range,
+		})
+		v.warningIdx++
+	}
+
+	for i, l := range v.logs {
+		log := VertexLog{
+			Message: l[v.logsOffset:],
+		}
+		if i == len(v.logs)-1 {
+			log.Partial = v.logsPartial
+		}
+		js.buf.Logs = append(js.buf.Logs, log)
+	}
+
+	if len(v.logs) > 0 {
+		if v.logsPartial {
+			v.logs = v.logs[len(v.logs)-1:]
+			v.logsOffset = len(v.logs[0])
+		} else {
+			v.logs = nil
+			v.logsOffset = 0
+		}
+	}
+
+	p.current = v.Digest
+	if v.isCompleted() && !isOpenStatus {
+		p.current = ""
+		v.count = 0
+
+		if v.Error != "" {
+			if strings.HasSuffix(v.Error, context.Canceled.Error()) {
+				js.buf.Vertex.Canceled = true
+			} else {
+				js.buf.Vertex.Error = v.Error
+			}
+		} else if v.Cached {
+			js.buf.Vertex.Cached = true
+		}
+		js.buf.Vertex.Completed = v.Completed
+	}
+	js.print(t.w)
+}
+
+// print will marshal the current status buffer and then reset it.
+// Does nothing if the status buffer contains no data.
+func (js *jsonPrinter) print(w io.Writer) {
+	if js.isEmpty() {
+		return
+	}
+	out, _ := json.Marshal(&js.buf)
```

Comment:
`json.NewEncoder(w); enc.SetIndent("", "  ");`

Reply:
Can you expand on this? My intention was for the stream to come out unformatted and tools like `jq` or something else could perform the formatting if desired. Adding an intent will increase the amount of memory that this progress method utilizes.

I also may be wrong about this, but I believe there isn't a difference between `json.NewEncoder(w).Encode(v)` and `json.Marshal(v)` except the former has an extra memory allocation. The internal implementation of the encoder uses the same internal methods as marshal.

Looking at the source, there is a difference between the two when indenting is used, but `MarshalIndent` is likely more performant unless you store the `json.Encoder` struct.

See the source code in the standard library here: https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/encoding/json/encode.go;l=179

In this one, it preallocates byte space for the indentation using a pretty simple metric. In contrast, here's the encoder variation: https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/encoding/json/stream.go;l=224

This reuses a buffer that's kept in the encoder. But, this will only have allocated space if there was a previous encoding that happened. Since this struct is created, used, and immediately discarded, I believe it just increases memory usage.

Reply:
`json.NewEncoder` is optional. We already have a writer, so I think it is clearer to use the writer-based API rather than allocate output into a byte slice and write/format it into writer then. I don't think there is a memory allocation difference(as afaic atm `Encode` does not do partial writes for error handling), or at least that is not meaningful. The indent is because JSON format is also human readable, not only machine-readable, and we shouldn't make it harder for users to understand it. I don't think the marshaling speed is relevant here.